atspi-connection/src/p2p.rs

```diff
@@ -0,0 +1,746 @@
+	async fn insert_well_known(
+		&self,
+		well_known_name: OwnedWellKnownName,
+		name_owner: OwnedUniqueName,
+		conn: &zbus::Connection,
+	) -> AtspiResult<()> {
+		let bus_name = BusName::WellKnown(well_known_name.inner().clone());
+		let address = bus_name.get_p2p_address(conn).await?;
+		let p2p_connection = Builder::address(address.clone())?.p2p().build().await?;
+
+		let peer = Peer {
+			unique_name: name_owner,
+			well_known_name: Some(well_known_name),
+			socket_address: address,
+			p2p_connection,
+		};
+
+		self.peers.lock().await.push(peer);
+		Ok(())
+	}
+
+	/// Removes a `Peer` with a well-known name from the list of peers.
+	async fn remove_well_known(
+		&self,
+		well_known_name: &OwnedWellKnownName,
+		name_owner: &OwnedUniqueName,
+	) -> AtspiResult<()> {
+		let mut peers = self.peers.lock().await;
+		peers.retain(|peer| {
+			!(peer.well_known_name() == Some(well_known_name) && peer.unique_name() == name_owner)
+		});
+		Ok(())
+	}
+
+	/// Update a `Peer` with a new owner of it's well-known name in the list of peers.
+	async fn update_well_known_owner(
+		&self,
+		well_known_name: OwnedWellKnownName,
+		old_name_owner: OwnedUniqueName,
+		new_name_owner: OwnedUniqueName,
+		conn: &zbus::Connection,
+	) -> AtspiResult<()> {
+		let address = BusName::from(new_name_owner.clone()).get_p2p_address(conn).await?;
+		let p2p_connection = Builder::address(address.clone())?.p2p().build().await?;
+
+		let peer = Peer {
+			unique_name: new_name_owner,
+			well_known_name: Some(well_known_name.clone()),
+			socket_address: address,
+			p2p_connection,
+		};
+
+		let mut peers = self.peers.lock().await;
+		if let Some(existing_peer) = peers.iter_mut().find(|p| {
+			p.well_known_name() == Some(&well_known_name) && p.unique_name() == &old_name_owner
+		}) {
+			*existing_peer = peer;
+		} else {
+			return Err(AtspiError::Owned(format!("Owner swap failed: well-known name {well_known_name} with owner: {old_name_owner} not found")));
+		}
+		Ok(())
+	}
+
+	/// Spawns a task which listens for  peer mutations.
+	///
+	/// This task listens for `NameOwnerChanged` signals and updates the list of peers accordingly.
+	///
+	/// # Async executor
+	/// This function uses the `async_executor::Executor` to spawn a task that listens for `NameAcquired` and `NameLost` signals on the `DBus`.
+	///
+	/// # Note
+	/// This function is called internally by `AccessibilityConnection::new()`.
+	pub(crate) fn spawn_peer_listener_task(
+		&self,
+		conn: &zbus::Connection,
+		dbus_proxy: DBusProxy<'_>,
+	) {
+		let executor = Executor::new();
+
+		executor
+			.spawn(async move {
+				let mut name_owner_changed_stream = match dbus_proxy.receive_name_owner_changed().await {
+					Ok(stream) => stream,
+					Err(_err) => {
+						#[cfg(feature = "tracing")]
+						warn!("Failed to get DBusProxy `NameOwnerChanged` stream: {}", _err);
+						return;
+					}
+				};
+
+				while let Some(name_owner_event) = name_owner_changed_stream.next().await {
```

Comment:
Great documentation here on how this all works. It's just a lot to read in one go.

It is _very_ deeply nested. It would be better to separate:

1. The spawned async task
2. The very long, nested match arms.

into their own functions.

Reply:
I've "separate concerns" here which mitigates complexity in the long match (to some degree). You may want to revisit since the last time.